core/src/main/scala/org/apache/spark/deploy/rest/SubmitRestProtocolMessage.scala

```diff
@@ -0,0 +1,201 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.rest
+
+import scala.collection.Map
+import scala.collection.JavaConversions._
+
+import org.json4s.jackson.JsonMethods._
+import org.json4s.JsonAST._
+
+import org.apache.spark.Logging
+import org.apache.spark.util.Utils
+
+/**
+ * A general message exchanged in the stable application submission REST protocol.
+ *
+ * The message is represented by a set of fields in the form of key value pairs.
+ * Each message must contain an ACTION field, which fully specifies the type of the message.
+ * For compatibility with older versions of Spark, existing fields must not be removed or
+ * modified, though new fields can be added as necessary.
+ */
+private[spark] abstract class SubmitRestProtocolMessage(
+    action: SubmitRestProtocolAction,
+    actionField: ActionField,
+    requiredFields: Seq[SubmitRestProtocolField]) {
+
+  // Maintain the insert order for converting to JSON later
+  private val fields = new java.util.LinkedHashMap[SubmitRestProtocolField, String]
+  val className = Utils.getFormattedClassName(this)
+
+  // Set the action field
+  fields.put(actionField, action.toString)
+
+  /** Return all fields currently set in this message. */
+  def getFields: Map[SubmitRestProtocolField, String] = fields.toMap
```

Comment:
I'm with @vanzin on this ... yes I'll agree "magical tools" often don't work as simply as advertised, but thats normally because you're trying to do something complicated, and its only more likely your home grown solution will also have an issue.  Its probably worth writing some unit tests to verify things work as expected, but you've got to do that with anything home grown in any case.

Even say it doesn't work well with scala case classes, and we should use plain java classes, thats still a lot better something totally homegrown.

Just as one tiny example of how this is worse than converting normal java objects: because you're trying to do manual type checking at runtime, you've converted a little typo like `message.setField(SUCCESS,"tru")` from a compile time error to a runtime error.